### PR TITLE
Convert all singleton lists in the MR APIs to embedded objects

### PR DESCRIPTION
**Pull request by ulucinar**
### Description of your changes

Depends on: https://github.com/crossplane/upjet/pull/387

Terraform configuration blocks, even if they have a MaxItems constraint of 1, are (almost) always generated as lists. We
  now generate the lists with a MaxItems constraint of 1 as embedded objects in our MR APIs.

This also helps when updating or patching via SSA the (previously list) objects. The merging strategy implemented by SSA requires configuration for associative lists and converting the singleton lists into embedded objects removes the configuration need.

The provider generates the converted embedded objects in the new `v1beta2` versions of the CRD APIs and registers upjet's identity converter and the singleton list converters to be invoked, in chain, by the CRD API conversion webhooks. This implies that the `v1beta1` versions stay intact, and old clients not aware of the new APIs should continue functioning as before (backwards-compatibility). Any clients willing to use the embedded objects-based APIs should be updated to use the `v1beta2` versions of the CRD APIs.

If a resource contains no singleton lists, then the `v1beta2` version is not generated for it.

We had to configure field name overrides for `HTTP{S}HealthCheck.compute` and `Version.dialogflowcx` so that the correspnding type names for the `spec.forProvider`, `spec.initParameter` and `state.atProvider` fields are not duplicated with the introduction of `v1beta2` packages.

We have also removed the SSA merge strategy configurations from the `v1beta2` versions of `Cluster.container` & `NodePool.container` resources because the related fields are no longer lists and have been converted to embedded objects in their `v1beta2` versions.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.
- [x] New version for the converted CRD APIs
- [x] Adjust the manually authored example manifests
- [x] API Conversion functions (invoked by the Conversion webhooks) to convert between the new & old API versions

### How has this code been tested

A successful uptest run with the converted APIs is observed here: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8723779942

Examples from successful uptest runs with conversion Webhooks enabled:
- https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8756638329/
- https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8756574997

**Uptest runs for available uptestable resources:**

*v1beta1 resources*
- `examples/administrativeunits/v1beta1/unit.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783425311
- `examples/app/v1beta1/roleassignment.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783436140
- `examples/applications/v1beta1/application.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783428288
- `examples/applications/v1beta1/certificate.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783429397
- `examples/applications/v1beta1/federatedidentitycredential.yaml` https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783592783
- `examples/applications/v1beta1/password.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783850825
- `examples/applications/v1beta1/preauthorized.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783430121
- `examples/conditionalaccess/v1beta1/accesspolicy.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783431498
- `examples/conditionalaccess/v1beta1/location.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783432163
- `examples/directoryroles/v1beta1/customdirectoryrole.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783432941
- `examples/directoryroles/v1beta1/roleassignment.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8782162727
- `examples/groups/v1beta1/group.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783599499
- `examples/groups/v1beta1/member.yaml`: *Uptest failed in the delete step (testhooks/delete-member.sh: permission denied), manual test successful.*
- `examples/policies/v1beta1/claimsmappingpolicy.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783602801
- `examples/serviceprincipaldelegated/v1beta1/permissiongrant.yaml`: *Uptest failed in the delete step (testhooks/delete-permissiongrant.sh: permission denied), manual test successful.*
- `examples/serviceprincipals/v1beta1/certificate.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8787919126
- `examples/serviceprincipals/v1beta1/claimsmappingpolicyassignment.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8784604890
- `examples/serviceprincipals/v1beta1/password.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8785038521
- `examples/serviceprincipals/v1beta1/principal.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8784021673
- `examples/synchronization/v1beta1/job.yaml`: *Uptest failed in the delete step (testhooks/delete-secret.sh: permission denied), manual test successful.*
- `examples/users/v1beta1/user.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783620282

*v1beta2 ressources*

- `examples/applications/v1beta2/application.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783430807
- `examples/conditionalaccess/v1beta2/accesspolicy.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783431498
- `examples/conditionalaccess/v1beta2/location.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783929325
- `examples/groups/v1beta2/group.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8783439239
- `examples/serviceprincipals/v1beta2/principal.yaml`: https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8784864220

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/location.yaml"

**Comment by ulucinar** [comment]
Hi @sergenyalcin,
Looks like the `status` subresource we are doing as part of the import tests (after scaling the provider deployment down) is now failing with the conversion webhooks:
https://github.com/crossplane-contrib/provider-upjet-azuread/actions/runs/8749171407/job/24010390302

image

I'll just disable the import tests for `v1beta2` (non-storage version) for now. They should be running fine for `v1beta1`, which does not require a conversion when updated and I think it should be fine to update the `status` (of `v1beta1`) when the provider pod is down.

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/location.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/location.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta1/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/location.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/location.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml,examples/conditionalaccess/v1beta2/location.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/location.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/groups/v1beta2/group.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/applications/v1beta1/application.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml,examples/conditionalaccess/v1beta2/location.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/applications/v1beta1/application.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/location.yaml,examples/applications/v1beta1/application.yaml,examples/groups/v1beta2/group.yaml"

File: Makefile [hunk 1]
```diff
@@ -53,12 +53,28 @@ GO_SUBDIRS += cmd internal apis
...
 
-KIND_VERSION = v0.15.0
-UP_VERSION = v0.16.1
+KIND_VERSION = v0.21.0
+UP_VERSION = v0.28.0
 UP_CHANNEL = stable
-UPTEST_VERSION = v0.7.0
+UPTEST_VERSION = v0.11.1
+UPTEST_LOCAL_VERSION = v0.12.0-2.g635fa3e
+UPTEST_LOCAL_CHANNEL = main
```
Comment (ulucinar) [inline Makefile:61]: Please see the comment below on the new `$(UPTEST_LOCAL)` target.

File: Makefile [hunk 2]
```diff
@@ -53,12 +53,28 @@ GO_SUBDIRS += cmd internal apis
...
+# uptest download and install
+UPTEST_LOCAL := $(TOOLS_HOST_DIR)/uptest-$(UPTEST_LOCAL_VERSION)
+
+$(UPTEST_LOCAL):
+	@$(INFO) installing uptest $(UPTEST_LOCAL)
+	@mkdir -p $(TOOLS_HOST_DIR)
+	@curl -fsSLo $(UPTEST_LOCAL) https://s3.us-west-2.amazonaws.com/crossplane.uptest.releases/$(UPTEST_LOCAL_CHANNEL)/$(UPTEST_LOCAL_VERSION)/bin/$(SAFEHOST_PLATFORM)/uptest || $(FAIL)
+	@chmod +x $(UPTEST_LOCAL)
+	@$(OK) installing uptest $(UPTEST_LOCAL)
+
```
Comment (ulucinar) [inline Makefile:77]: This is a temporary workaround till https://github.com/upbound/build/pull/255 is merged.

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml,examples/conditionalaccess/v1beta2/location.yaml,examples/applications/v1beta1/application.yaml,examples/groups/v1beta2/group.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml,examples/conditionalaccess/v1beta2/location.yaml,examples/applications/v1beta1/application.yaml,examples/groups/v1beta2/group.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/administrativeunits/v1beta1/unit.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/applications/v1beta1/application.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/applications/v1beta1/certificate.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/applications/v1beta1/preauthorized.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/applications/v1beta2/application.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/conditionalaccess/v1beta1/accesspolicy.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/conditionalaccess/v1beta1/location.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/directoryroles/v1beta1/roleassignment.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/groups/v1beta2/group.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/directoryroles/v1beta1/roleassignment.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/administrativeunits/v1beta1/unit.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/applications/v1beta1/application.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/applications/v1beta1/certificate.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/applications/v1beta1/preauthorized.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/applications/v1beta2/application.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/conditionalaccess/v1beta1/accesspolicy.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/conditionalaccess/v1beta1/location.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/directoryroles/v1beta1/customdirectoryrole.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/app/v1beta1/roleassignment.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/groups/v1beta2/group.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/applications/v1beta1/federatedidentitycredential.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/groups/v1beta1/group.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/policies/v1beta1/claimsmappingpolicy.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/serviceprincipaldelegated/v1beta1/permissiongrant.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/synchronization/v1beta1/job.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/users/v1beta1/user.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/applications/v1beta1/password.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/groups/v1beta1/member.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/serviceprincipals/v1beta1/certificate.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/conditionalaccess/v1beta2/location.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/serviceprincipals/v1beta1/principal.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/groups/v1beta1/member.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/serviceprincipals/v1beta1/claimsmappingpolicyassignment.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/groups/v1beta1/member.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/serviceprincipals/v1beta1/password.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/serviceprincipals/v1beta1/certificate.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/applications/v1beta1/application.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/applications/v1beta1/application.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/groups/v1beta1/member.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/serviceprincipaldelegated/v1beta1/permissiongrant.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/synchronization/v1beta1/job.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/groups/v1beta1/member.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/serviceprincipaldelegated/v1beta1/permissiongrant.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/synchronization/v1beta1/job.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/groups/v1beta1/member.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/conditionalaccess/v1beta2/accesspolicy.yaml"

**Comment by ulucinar** [comment]
/test-examples="examples/groups/v1beta1/member.yaml"

**Comment by turkenf** [comment]
**Upgrada test**
- Installed the `package: xpkg.upbound.io/upbound/provider-azuread:v1.1.0`
-  Applied the [examples/conditionalaccess/v1beta1/accesspolicy.yaml](https://github.com/crossplane-contrib/provider-upjet-azuread/blob/main/examples/conditionalaccess/v1beta1/accesspolicy.yaml)
- We can see the singleton list in `get -o yaml` output:
```
    conditions:
    - applications:
      - includedApplications:
        - All
```
- Edited the package with `package: xpkg.upbound.io/upbound/provider-azuread:v1.2.0-rc.0.16.g14395d1`
```
NAME               INSTALLED   HEALTHY   PACKAGE                                                            AGE
provider-azuread   True        True      xpkg.upbound.io/upbound/provider-azuread:v1.2.0-rc.0.16.g14395d1   11m
```
- and we can see the following output from `get -o yaml`
     - apiVersion:
     ```
     apiVersion: conditionalaccess.azuread.upbound.io/v1beta2
     ```
     - embedded objects:
     - conditions:

**Pull request merged**